spritecontrol.cpp

```diff
@@ -391,6 +392,247 @@ std::map<PatternName, GlobalPlayer::MoveFunction> GlobalPlayer::move_functions {
+		std::map<Sprite *, Point> alignment_velocity_changes;
+		std::map<Sprite *, Point> cohesion_velocity_changes;
+		std::map<Sprite *, Point> edge_push_velocity_changes;
+
+		for (Sprite *current_sprite : *sprites) {
+			Point current_final = Point(current_sprite->final<X>(), current_sprite->final<Y>());
+			size_t sprites_seen = 1;
+			double average_angle = get_angle(velocity[current_sprite->id()]);
+			Point average_pos = Point(current_final);
+
+			for (Sprite *other_sprite : *sprites) {		// dejil... i am sorry...
+				if (current_sprite == other_sprite) {
+					continue;
+				}
+
+				Point other_final = Point(other_sprite->final<X>(), other_sprite->final<Y>());
+				Point diff = current_final - other_final;
+				get<Y>(diff) *= STRETCH_RATIO;
+
+				double dist = get_vector_magnitude(diff);
+				
+				if (dist > vision_range[current_sprite->id()]) {
+					continue;
+				}
+
+				double current_angle = get_angle(velocity[current_sprite->id()]);
+				double relative_angle = get_angle(diff);
+				relative_angle = wrap_angle(relative_angle - current_angle);	// 180 deg: straight ahead; 0 deg: straight behind (assuming i'm not bad at math)
+
+				if (abs(relative_angle) >= blind_angle[current_sprite->id()]) {
+					sprites_seen++;
+					double other_angle = get_angle(velocity[other_sprite->id()]);
+					average_angle += other_angle;
+					average_pos += other_final;
+				}
+
+				if (dist < desired_separation[current_sprite->id()]) {
+					diff /= dist * dist / desired_separation[current_sprite->id()];
+					separation_velocity_changes[current_sprite] = diff;
+				}
+			}
+
+			if (sprites_seen > 1) {
+				average_angle /= sprites_seen;
+				average_pos /= (double) sprites_seen;
+
+				// i sure hope the angles i'm calculating line up with the angles used for movement
+				alignment_velocity_changes[current_sprite] = Point(cos(average_angle), sin(average_angle));
+
+				Point relative_average_pos = average_pos - Point(current_sprite->final<X>(), current_sprite->final<Y>());
+				cohesion_velocity_changes[current_sprite] = normalize_vector(relative_average_pos);
+			}
+
+			Point edge_push = Point(0.0, 0.0);
+
+			if (-1.0 + EDGE_PUSH_X_RADIUS > get<X>(current_final)) {
+				get<X>(edge_push) += ((-1.0 + EDGE_PUSH_X_RADIUS) - get<X>(current_final)) / EDGE_PUSH_X_RADIUS;
+			} else if (1.0 - EDGE_PUSH_X_RADIUS < get<X>(current_final)) {
+				get<X>(edge_push) += ((1.0 - EDGE_PUSH_X_RADIUS) - get<X>(current_final)) / EDGE_PUSH_X_RADIUS;
+			}
+
+			if (-1.0 + EDGE_PUSH_X_RADIUS > get<Y>(current_final)) {
+				get<Y>(edge_push) += ((-1.0 + EDGE_PUSH_X_RADIUS) - get<Y>(current_final)) / EDGE_PUSH_X_RADIUS;
+			} else if (1.0 - EDGE_PUSH_X_RADIUS < get<Y>(current_final)) {
+				get<Y>(edge_push) += ((1.0 - EDGE_PUSH_X_RADIUS) - get<Y>(current_final)) / EDGE_PUSH_X_RADIUS;
+			}
+
+			if (edge_push != Point(0.0, 0.0)) {
+				get<X>(edge_push) = pow(abs(get<X>(edge_push)), 1.0/2.5) * (get<X>(edge_push) < 0.0 ? -1.0 : 1.0);
```

Comment:
Random thought: while having our Point type just be a std::pair is nice, I really have grown to hate the ergonomics of this `get<X>` crap, and really wish we could just do point.x and point.y instead. That is very possible but it's the scope of another PR.